File: crates/kanwise/src/agent/worktree.rs
use anyhow::{Context, Result};
use std::path::{Path, PathBuf};
use std::process::Command;

/// Create a git worktree for an agent session.
/// Returns the path to the worktree directory.
pub fn create_worktree(
    repo_dir: &Path,
    session_id: &str,
    branch_name: &str,
) -> Result<PathBuf> {
    let worktree_dir = repo_dir.join(".worktrees").join(session_id);

    // Ensure .worktrees is in .gitignore
    ensure_gitignore(repo_dir)?;

    let output = Command::new("git")
        .current_dir(repo_dir)
        .args([
            "worktree",
            "add",
            &worktree_dir.display().to_string(),
            "-b",
            branch_name,
        ])
        .output()
        .context("failed to run git worktree add")?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        anyhow::bail!("git worktree add failed: {stderr}");
    }

    Ok(worktree_dir)
}

/// Remove a git worktree and its branch.
pub fn cleanup_worktree(repo_dir: &Path, session_id: &str, branch_name: &str) -> Result<()> {
    let worktree_dir = repo_dir.join(".worktrees").join(session_id);

    if worktree_dir.exists() {
        let _ = Command::new("git")
            .current_dir(repo_dir)
            .args([
                "worktree",
                "remove",
                "--force",
                &worktree_dir.display().to_string(),
            ])
            .output();
    }

    // Clean up the branch
    let _ = Command::new("git")
        .current_dir(repo_dir)
        .args(["branch", "-D", branch_name])
        .output();

    Ok(())
}

fn ensure_gitignore(repo_dir: &Path) -> Result<()> {
    let gitignore = repo_dir.join(".gitignore");
    if gitignore.exists() {
        let content = std::fs::read_to_string(&gitignore)?;
        if content.contains(".worktrees") {
            return Ok(());
        }
        let mut content = content;
        if !content.ends_with('\n') {
            content.push('\n');
        }
        content.push_str(".worktrees/\n");
        std::fs::write(&gitignore, content)?;
    } else {
        std::fs::write(&gitignore, ".worktrees/\n")?;
    }
    Ok(())
}

/// Generate a branch name from task and session IDs.
pub fn branch_name(task_id: &str, session_id: &str) -> String {
    let task_short = &task_id[..std::cmp::min(8, task_id.len())];
    let session_short = &session_id[..std::cmp::min(8, session_id.len())];
    format!("agent/{task_short}-{session_short}")
}

File: agent/src/worktree.ts
import { execFile } from "child_process";
import { promisify } from "util";
import fs from "fs/promises";
import path from "path";

const exec = promisify(execFile);

export function branchName(taskId: string, sessionId: string): string {
  const taskShort = taskId.slice(0, 8);
  const sessShort = sessionId.slice(0, 8);
  return `agent/${taskShort}-${sessShort}`;
}

export async function createWorktree(
  repoDir: string,
  sessionId: string,
  branch: string
): Promise<string> {
  const wtDir = path.join(repoDir, ".worktrees", sessionId);
  await ensureGitignore(repoDir);
  await exec("git", ["worktree", "add", wtDir, "-b", branch], {
    cwd: repoDir,
  });
  return wtDir;
}

export async function cleanupWorktree(
  repoDir: string,
  sessionId: string,
  branch: string
): Promise<void> {
  const wtDir = path.join(repoDir, ".worktrees", sessionId);
  try {
    await exec("git", ["worktree", "remove", "--force", wtDir], {
      cwd: repoDir,
    });
  } catch {
    // best-effort
  }
  try {
    await exec("git", ["branch", "-D", branch], { cwd: repoDir });
  } catch {
    // best-effort
  }
}

export async function cleanupOrphanedWorktrees(
  repoDir: string
): Promise<void> {
  const wtBase = path.join(repoDir, ".worktrees");
  try {
    const entries = await fs.readdir(wtBase);
    for (const entry of entries) {
      try {
        await exec("git", ["worktree", "remove", "--force", path.join(wtBase, entry)], {
          cwd: repoDir,
        });
      } catch {
        // skip
      }
    }
  } catch {
    // .worktrees/ doesn't exist — nothing to clean
  }
}

async function ensureGitignore(repoDir: string): Promise<void> {
  const gitignorePath = path.join(repoDir, ".gitignore");
  try {
    const content = await fs.readFile(gitignorePath, "utf-8");
    if (content.includes(".worktrees/")) return;
    await fs.appendFile(gitignorePath, "\n.worktrees/\n");
  } catch {
    await fs.writeFile(gitignorePath, ".worktrees/\n");
  }
}
